File: python/proxy/proxy1/DatabaseServiceProxy.py
from DatabaseService import DatabaseService
from RealDatabaseService import RealDatabaseService

# Proxy que controla o acesso ao serviço real
class DatabaseServiceProxy(DatabaseService):
    def __init__(self):
        self._real_database_service = None
        self._cache_valid = False

    def request_data(self):
        if self._cache_valid:
            print("Usando dados do cache...")
        else:
            if self._real_database_service is None:
                self._real_database_service = RealDatabaseService()
            self._real_database_service.request_data()
            self._cache_valid = True
            print("Dados armazenados no cache.")

File: java/proxy/proxy/src/main/java/edu/uea/DatabaseServiceProxy.java
package edu.uea;

// Proxy que controla o acesso ao serviço real
public class DatabaseServiceProxy implements DatabaseService {
    private RealDatabaseService realDatabaseService;
    private boolean cacheValid = false;

    @Override
    public void requestData() {
        if (cacheValid) {
            System.out.println("Usando dados do cache...");
        } else {
            if (realDatabaseService == null) {
                realDatabaseService = new RealDatabaseService();
            }
            realDatabaseService.requestData();
            cacheValid = true;
            System.out.println("Dados armazenados no cache.");
        }
    }

}
